src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs

```diff
@@ -0,0 +1,367 @@
+                    }
+                    return true;
+                });
+            }
+
+            if (dateRequest != ActivityDataRequest.None)
+            {
+                activity = Activity.CreateAndStart(this, name, kind, parentId, context, tags, links, startTime, dateRequest);
+                listeners.EnumWithAction(listener => {
+                    var activityStarted = listener.ActivityStarted;
+                    if (activityStarted != null)
+                    {
+                        activityStarted(activity);
+                    }
+                });
+            }
+
+            return activity;
+        }
+
+        /// <summary>
+        /// Dispose the ActivitySource object and remove the current instance from the global list. empty the listeners list too.
+        /// </summary>
+        public void Dispose()
+        {
+            _listeners = null;
+            s_activeSources.Remove(this);
+        }
+
+        /// <summary>
+        /// Add a listener to the <see cref="Activity"/> starting and stopping events.
+        /// </summary>
+        /// <param name="listener"> The <see cref="ActivityListener"/> object to use for listeneing to the <see cref="Activity"/> events.</param>
+        public static void AddActivityListener(ActivityListener listener)
+        {
+            if (listener == null)
+            {
+                throw new ArgumentNullException(nameof(listener));
+            }
+
+            if (s_allListeners.AddIfNotExist(listener))
+            {
+                s_activeSources.EnumWithAction(source => {
+                    var shouldListenTo = listener.ShouldListenTo;
+                    if (shouldListenTo != null && shouldListenTo(source))
+                    {
+                        source.AddListener(listener);
+                    }
+                });
+            }
+        }
+
+        internal void AddListener(ActivityListener listener)
+        {
+            if (_listeners == null)
+            {
+                Interlocked.CompareExchange(ref _listeners, new SynchronizedList<ActivityListener>(), null);
+            }
+
+            _listeners.AddIfNotExist(listener);
+        }
+
+        internal static void DetachListener(ActivityListener listener)
+        {
+            s_allListeners.Remove(listener);
+
+            s_activeSources.EnumWithAction(source => {
+                var listeners = source._listeners;
+                listeners?.Remove(listener);
+            });
+        }
+
+        internal void NotifyActivityStart(Activity activity)
+        {
+            Debug.Assert(activity != null);
+
+            // _listeners can get assigned to null in Dispose.
+            SynchronizedList<ActivityListener>? listeners = _listeners;
+            if (listeners != null &&  listeners.Count > 0)
+            {
+                listeners.EnumWithAction(listener => {
+                    var activityStarted = listener.ActivityStarted;
```

Comment:
Nit: I believe you can compact 5 lines => 1 line by writing listener.ActivityStarted?.Invoke(activity)

Reply:
Initially, I though x?.y is not thread-safe. but after checking the compiler and asking some people I found the compiler making a copy which makes it thread-safe. I have applied what you have suggested.